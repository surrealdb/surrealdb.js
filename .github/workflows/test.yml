name: test

on:
  push:
  pull_request:

jobs:
  test:
    runs-on: ubuntu-latest

    steps:
      - name: Setup node.js
        uses: actions/setup-node@v3
        with:
          node-version: 18
      - name: Setup deno
        uses: denoland/setup-deno@v1
        with:
          deno-version: v1.x
      - name: Setup bun
        uses: antongolub/action-setup-bun@v1
        with:
          bun-version: 0.1.13

      - uses: actions/checkout@v3

      - name: Build NPM package
        run: deno task build

      - name: Start SurrealDB
        run: sh ./test/e2e/scripts/docker-restart.sh

      - name: Run Deno
        run: deno test -A --trace-ops ./test/e2e/deno.js
      - name: Run Node
        run: node ./test/e2e/node.js
      - name: Run Bun
        run: bun run ./test/e2e/bun.js

      - name: Stop SurrealDB
        run:  sh ./test/e2e/scripts/docker-stop.sh
