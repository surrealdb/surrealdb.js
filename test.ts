import { RecordId } from "./src/library/data/recordid.ts";
import { Duration } from "./src/library/data/duration.ts";
import { UUID } from "./src/library/data/uuid.ts";
import { Decimal } from "./src/library/data/decimal.ts";
import { Surreal } from "./src/index.ts";
import { encode } from "./src/library/data/cbor.ts";
import { GeometryCollection, GeometryLine, GeometryPoint } from "./src/library/data/geometry.ts";
// import { recordId } from "./src/library/orm/types.ts";

const surreal = new Surreal();
await surreal.connect('ws://127.0.0.1:8000/rpc');
await surreal.use({
	namespace: 'test',
	database: 'test',
});

// console.log(
// 	await surreal.query<[unknown[]]>(
// 		"[$d, $r, $c, $u, $dur, $none, NONE, $a, <string> $a, $nan, $geo]",
// 		{
// 			d: new Date(),
// 			r: new RecordId('person', 'tobie'),
// 			c: new RecordId('recording', [
// 				new Date(),
// 				'London',
// 				new Uuid(),
// 				{
// 					temparature: new Decimal("10.00003")
// 				}
// 			]),
// 			u: new Uuid(),
// 			dur: new Duration('1d20m'),
// 			none: undefined,
// 			a: 9223372036854775807n,
// 			nan: NaN,
// 			geo: new GeometryCollection([
// 				new GeometryPoint([1, 2]),
// 				new GeometryLine([ new GeometryPoint([1, 2]), new GeometryPoint([3, 4]) ])
// 			])
// 		}
// 	)
// )

const uuid = new Duration("2w6d100ms");
console.log(
	456,
	await surreal.query<[UUID, string]>(
		"$uuid; <string> $uuid",
		{
			uuid
		}
	).then(([u, a]) => [u.toString(), a])
)

// function bufferToHex (buffer: ArrayBuffer) {
//     return [...new Uint8Array (buffer)]
//         .map (b => b.toString (16).padStart (2, "0"))
//         .join ("");
// }

// console.log(bufferToHex(encode(new RecordId('recording', [
// 	new Date(),
// 	'London',
// 	new Uuid(),
// 	{
// 		temparature: new Decimal("10.00003")
// 	}
// ]))))

// const rid = recordId('test');
// console.log(rid.parse(new RecordId('tesst', 123) ))

// console.log()
