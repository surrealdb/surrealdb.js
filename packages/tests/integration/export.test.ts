import { beforeAll, describe, expect, test } from "bun:test";
import { compareVersions } from "compare-versions";
import { fetchVersion, setupServer } from "./__helpers__";

const { createSurreal } = await setupServer();

beforeAll(async () => {
    const surreal = await createSurreal();

    await surreal.query(/* surql */ `
		CREATE foo:1 CONTENT { hello: "world" };
		CREATE bar:1 CONTENT { hello: "world" };
		DEFINE FUNCTION fn::foo() { RETURN "bar"; };
	`);
});

describe("export", async () => {
    const surreal = await createSurreal();
    const version = await fetchVersion(surreal);
    const is3x = compareVersions(version, "3.0.0-alpha.1") >= 0;
    const is2x = compareVersions(version, "3.0.0-alpha.1") < 0;

    test.if(is2x)("basic 2.x", async () => {
        const res = await surreal.export();

        expect(res).toMatchSnapshot();
    });

    test.if(is3x)("basic 3.x", async () => {
        const res = await surreal.export();

        expect(res).toMatchSnapshot();
    });

    test.if(is2x)("filter tables 2.x", async () => {
        const res = await surreal.export({
            tables: ["foo"],
        });

        expect(res).toMatchSnapshot();
    });

    test.if(is3x)("filter tables 3.x", async () => {
        const res = await surreal.export({
            tables: ["foo"],
        });

        expect(res).toMatchSnapshot();
    });

    test.if(is2x)("filter functions 2.x", async () => {
        const res = await surreal.export({
            functions: true,
            tables: false,
        });

        expect(res).toMatchSnapshot();
    });

    test.if(is3x)("filter functions 3.x", async () => {
        const res = await surreal.export({
            functions: true,
            tables: false,
        });

        expect(res).toMatchSnapshot();
    });
});
