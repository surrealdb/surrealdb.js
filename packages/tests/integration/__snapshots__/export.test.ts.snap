// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`export basic 2.x 1`] = `
"-- ------------------------------
-- OPTION
-- ------------------------------

OPTION IMPORT;

-- ------------------------------
-- FUNCTIONS
-- ------------------------------

DEFINE FUNCTION fn::foo() { RETURN 'bar'; } PERMISSIONS FULL;

-- ------------------------------
-- TABLE: bar
-- ------------------------------

DEFINE TABLE bar TYPE ANY SCHEMALESS PERMISSIONS NONE;




-- ------------------------------
-- TABLE DATA: bar
-- ------------------------------

INSERT [ { hello: 'world', id: bar:1 } ];

-- ------------------------------
-- TABLE: foo
-- ------------------------------

DEFINE TABLE foo TYPE ANY SCHEMALESS PERMISSIONS NONE;




-- ------------------------------
-- TABLE DATA: foo
-- ------------------------------

INSERT [ { hello: 'world', id: foo:1 } ];

"
`;

exports[`export basic 3.x 1`] = `
"-- ------------------------------
-- OPTION
-- ------------------------------

OPTION IMPORT;

-- ------------------------------
-- FUNCTIONS
-- ------------------------------

DEFINE FUNCTION fn::foo() { RETURN 'bar' } PERMISSIONS FULL;

-- ------------------------------
-- TABLE: bar
-- ------------------------------

DEFINE TABLE bar TYPE ANY SCHEMALESS PERMISSIONS NONE;




-- ------------------------------
-- TABLE DATA: bar
-- ------------------------------

INSERT [ { hello: 'world', id: bar:1 } ];

-- ------------------------------
-- TABLE: foo
-- ------------------------------

DEFINE TABLE foo TYPE ANY SCHEMALESS PERMISSIONS NONE;




-- ------------------------------
-- TABLE DATA: foo
-- ------------------------------

INSERT [ { hello: 'world', id: foo:1 } ];

"
`;

exports[`export filter tables 2.x 1`] = `
"-- ------------------------------
-- OPTION
-- ------------------------------

OPTION IMPORT;

-- ------------------------------
-- FUNCTIONS
-- ------------------------------

DEFINE FUNCTION fn::foo() { RETURN 'bar'; } PERMISSIONS FULL;

-- ------------------------------
-- TABLE: foo
-- ------------------------------

DEFINE TABLE foo TYPE ANY SCHEMALESS PERMISSIONS NONE;




-- ------------------------------
-- TABLE DATA: foo
-- ------------------------------

INSERT [ { hello: 'world', id: foo:1 } ];

"
`;

exports[`export filter tables 3.x 1`] = `
"-- ------------------------------
-- OPTION
-- ------------------------------

OPTION IMPORT;

-- ------------------------------
-- FUNCTIONS
-- ------------------------------

DEFINE FUNCTION fn::foo() { RETURN 'bar' } PERMISSIONS FULL;

-- ------------------------------
-- TABLE: foo
-- ------------------------------

DEFINE TABLE foo TYPE ANY SCHEMALESS PERMISSIONS NONE;




-- ------------------------------
-- TABLE DATA: foo
-- ------------------------------

INSERT [ { hello: 'world', id: foo:1 } ];

"
`;

exports[`export filter functions 2.x 1`] = `
"-- ------------------------------
-- OPTION
-- ------------------------------

OPTION IMPORT;

-- ------------------------------
-- FUNCTIONS
-- ------------------------------

DEFINE FUNCTION fn::foo() { RETURN 'bar'; } PERMISSIONS FULL;

"
`;

exports[`export filter functions 3.x 1`] = `
"-- ------------------------------
-- OPTION
-- ------------------------------

OPTION IMPORT;

-- ------------------------------
-- FUNCTIONS
-- ------------------------------

DEFINE FUNCTION fn::foo() { RETURN 'bar' } PERMISSIONS FULL;

"
`;
